### ISSUE TITLE
grammar for "closepath can substitute for final coordinate" is broken

### ISSUE THREAD

**Issue by tshinnic** [w3c/svgwg#325]
Change https://github.com/w3c/svgwg/commit/7cce3e3f80d703660b95a3d3dc13529d9c417932 made many updates to the grammar, also including [allowing a 'Z' closepath command to substitute for the final coordinate pair](https://www.w3.org/TR/SVG2/paths.html#PathDataClosePathCommand) of a curve/arc command.  

However, in addition to accidentally dropping scientific notation, I believe that change doesn't create a correct grammar to support the intention of the "closepath replaces coordinates" update.

For instance, the `curveto` command has arguments "(x1 y1 x2 y2 x y)+".  The grammar has
```
curveto::=
    ("C"|"c") wsp* (curveto_coordinate_sequence | (coordinate_pair_sequence? closepath))

curveto_coordinate_sequence::=
    coordinate_pair_triplet
    | (coordinate_pair_triplet comma_wsp? curveto_coordinate_sequence)

coordinate_pair_triplet::=
    coordinate_pair comma_wsp? coordinate_pair comma_wsp? coordinate_pair

coordinate_pair_sequence::=
    coordinate_pair | (coordinate_pair comma_wsp? coordinate_pair_sequence)
```
Example data strings for `curveto` can be found in test `paths-data-19-f.svg`
```
  d="M50 150 C50 50 200 50 200 150 C200 50 350 50 350 150"
  d="M50 150 C50 50 200 50 200 150 200 50 350 50 350 150" 
  d="M50, 200 c0,-100 150,-100 150,0 c0,-100 150,-100 150,0"
  d="M50, 200 c0,-100 150,-100 150,0 0,-100 150,-100 150,0"
```
showing separate curve commands and single commands coding for multiple curves.  Note that `C50 50  200 50  200 150   200 50  350 50  350 150` draws two curves and ending at (350,150).  It has two coordinate pair triples.

My understanding of the `closepath` change is that the *last* coordinate, the end point for the *last* curve coded, can be omitted if followed directly by a closepath command.  The last curve's end point is then the initial point of the segment.

But if I'm reading the grammar correctly, one cannot specify `d="C50 50 200 50 200 150  200 50 350 50 Z"`, omitting the previous final "350 150".  Here there would be a triple pair, followed by a double pair.

The grammar would seem to insist on any number of triple pairs, or any number of double pairs followed by a closepath, e.g. `d="... C200 50 350 50 Z"`.  The command above, a triple pair and a double pair, would be invalid for this grammar.  

Note that the formulation of `elliptical_arc` employing `elliptical_arc_closing_argument` *seems* to be a correct recipe, though offputting.

The `curveto`, `smooth_curveto`, `quadratic_bezier` **grammars look wrong**.  The `smooth_quadratic_bezier` and `elliptical_arc` grammars might be correct.  All this deserves another hard look.

**Comment by AmeliaBR** [comment]
Thanks @tshinnic .  I should have added a warning on the WPT issue that there were other errors in the published grammar which have now been corrected in the [Editor's Draft version of the grammar](https://svgwg.org/svg2-draft/paths.html#PathDataBNF).

Are you able to review that version & confirm if it has corrected all the issues you discovered?

**Comment by tshinnic** [comment]
That looks very much like the grammar I was referencing and looks unhappily like the same problems.

What I'm going to try to do is translate that grammar to something testable, using as basis the project https://github.com/hughsk/svg-path-parser/ which uses [peg.js](https://pegjs.org/) underneath.  This is a product of Gavin Kistner (aka @Phrogz) working against the SVG 1.1 grammar.  I started hassling him there and ended up here.  ([just deserts](https://en.wikipedia.org/wiki/Desert_(philosophy)) for me?)

I'd rather have something more like concrete reasons (and demonstrable) for complaint than just "it looks wrong".

Question: Is there a sense of what the resolution for w3c/svgwg/issues/286 will be?  I could re-introduce `floating-point-constant` into the grammar and perhaps be able to show necessity vis-a-vis extant path strings?

**Comment by AmeliaBR** [comment]
I will say this much: If there are any changes in the path grammar (relative to [SVG 1.1](https://www.w3.org/TR/SVG11/paths.html#PathDataBNF)) that do not have matching changes in the prose, they were probably un-intentional.  Many of the errors probably come from trying to change from BNF to EBNF, to reduce the number of different syntaxes used throughout the SVG spec.

BTW, the original commit was https://github.com/w3c/svgwg/commit/23bebc27ec3f3bbd6f3428fc5df2ff2cc9800d9d by @nikosandronikos .  The one referenced above was simply copying & pasting between the SVG 2 Paths chapter and the SVG Paths Level 3 module.

@tshinnic 
If you are able to draft a correct EBNF that actually matches the prose, that would be a huge benefit.  Unfortunately, the SVG WG does not technically exist right now, so there won't be an updated W3C draft any time soon.  But some of us still have write permissions on the repo, so we could review a pull request and update the Editor's Draft (which is automatically built from the master folder of the master branch of this repo).

**Comment by nikosandronikos** [comment]
> But if I'm reading the grammar correctly, one cannot specify d="C50 50 200 50 200 150 200 50 350 50 Z", omitting the previous final "350 150". Here there would be a triple pair, followed by a double pair.

That path is invalid because there's no intial moveto.
The curve commands should be accepted by the
''' (coordinate_pair_sequence? closepath)'''

part of
'''
curveto::=
    ("C"|"c") wsp* (curveto_coordinate_sequence | (coordinate_pair_sequence? closepath))
'''
It's five sequences of '''coordinate_pair_sequence'''. The presence of the '''closepath''' forces it to be consumed by the second of the options.

When writing the grammar, I tested with:
https://services.w3.org/yacker/uploads/svg2b?lang=perl&text=M+10%2C10+C50%2C50+200%2C50+200%2C150+200%2C50+350+50+Z&action=validate+text
And this accepts that string.
_Note: For some reason I had to change the svg_path definition of that grammar to get it accepting anything. I'm not sure why and don't have much time to look into it..._

Personally I think the grammar should be removed from the spec. It's error prone without continuous testing and the prose is much easier to understand - the rules are nowhere near as complicated as the grammar makes them seem. I wish I'd argued for removal back in the day.

**Comment by nikosandronikos** [comment]
> My understanding of the closepath change is that the last coordinate, the end point for the last curve coded, can be omitted if followed directly by a closepath command. The last curve's end point is then the initial point of the segment.

You can omit multiple co-ordinates and any omitted co-ordinate is given the same value.

**Comment by boggydigital** [comment]
Not blocking updated 2.0 CR publication - assigning 2.0 Recommendation milestone to clean this up before 2.0 REC

**Comment by dirkschulze** [comment]
At least the path data EBNF should match the SVG 1.1 for the most part with a few modifications for scientific numbers. Closing the issue. Please reopen if there are issues with the EBNF or the prose text still. Even better, open individual issues for each observed issue.

**Comment by dirkschulze** [comment]
Fixed by #697.
